File: longestValidParentheses/fastestSolution.cpp
class Solution {
public:
	int longestValidParentheses(string s) {
		int n = s.size();
		vector<int> dp(n, 0);

		int open = 0, ans = 0;
		for (int i = 0; i < n; ++i) {
			if (s[i] == '(') {
				++open;
				continue;
			}

			if (open == 0) continue;
			dp[i] = dp[i - 1] + 2;
			if (i - dp[i] >= 0) dp[i] += dp[i - dp[i]];
			--open;
			ans = max(ans, dp[i]);
		}

		return ans;
	}
}

File: maxProductSubarray/fastest_solution.py
# https://leetcode.com/problems/maximum-product-subarray/

import math
from typing import List


class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        n = len(nums)
        if n == 1:
            return nums[0]
        res = math.prod(nums)
        if res > 0:
            return res
        max_till_now = min_till_now = 1
        for num in nums:
            prod_max = max_till_now * num
            prod_min = min_till_now * num

            max_till_now = max(prod_max, prod_min, num)
            min_till_now = min(prod_max, prod_min, num)
            res = max(res, max_till_now)
        return res
